### PR TITLE
Allow custom `const x = boolean` exports to be read from plugins

### PR DESCRIPTION
**Pull request by ottomated**
## Changes

- Any const boolean exports on a page can now be accessed on `RouteData`. This allows plugins to define custom per-page options.

## Testing

- Adds a new `custom-page-options` test

## Docs

- This should probably be added to plugin docs
/cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

File: packages/astro/src/vite-plugin-scanner/scan.ts
Comment (ematipico) [inline packages/astro/src/vite-plugin-scanner/scan.ts:57]: This `6` should be saved into a top-level `const` with a name that explains **why** 6 (maybe with a comment);

File: packages/astro/src/vite-plugin-scanner/scan.ts
Comment (ematipico) [inline packages/astro/src/vite-plugin-scanner/scan.ts:63]: This check won't work for carriage returns, e.g. `\r`

Also, we can't assume there's a semicolon.

File: packages/astro/test/fixtures/custom-page-options/src/pages/index.astro
Comment (ematipico) [inline packages/astro/test/fixtures/custom-page-options/src/pages/index.astro:5]: We should add more cases:
- no semicolons
- more spaces between `=` and the value;

File: .changeset/brown-tips-drop.md
Comment (ematipico) [inline .changeset/brown-tips-drop.md:5]: I would call it `externalOptions`. It's more sound and communicates that these options **don't** from Astro. Or `pluginOptions`.

Comment (matthewp) [reply .changeset/brown-tips-drop.md:5]: I don't think we should differentiate. If we accept this feature I think it should include all boolean options, including `prerender`.

The reason is, if this field is just for non-Astro defined booleans, that would mean that if Astro adds any new booleans itself, that becomes a breaking change. However if they are all under `customOptions` then there is no breaking change.

File: packages/astro/src/vite-plugin-scanner/scan.ts
Comment (ottomated) [reply packages/astro/src/vite-plugin-scanner/scan.ts:63]: - In the case of a carriage return, this will return `"value"\r` and the carriage return will be trimmed off on line 64
- It doesn't assume there's a semicolon – it splits off at either the first semicolon OR the first newline

Comment (ottomated) [reply packages/astro/src/vite-plugin-scanner/scan.ts:63]: Added a test case for this and it passes

File: packages/astro/src/vite-plugin-scanner/scan.ts
Comment (ottomated) [reply packages/astro/src/vite-plugin-scanner/scan.ts:53]: This only occurs if there is no local variable being exported - i.e. `export default 1`. That's impossible to do in a `.astro` file because it is already exporting the component as default.

File: .changeset/brown-tips-drop.md
Comment (matthewp) [reply .changeset/brown-tips-drop.md:5]: Another option here is that custom options must have some prefix we define. Similar to how HTML has the `data-` attributes. If they were required to be `customFoo`, for example, we could ensure that Astro never accidentally breaks an integration defined option.

File: packages/astro/src/vite-plugin-scanner/scan.ts
Comment (ottomated) [reply packages/astro/src/vite-plugin-scanner/scan.ts]: added an inline comment, don't think it needs to be top-level for this case

File: .changeset/brown-tips-drop.md
Comment (ottomated) [reply .changeset/brown-tips-drop.md:5]: I like that idea but `custom` as a prefix feels a bit bulky to me

**Review by natemoo-re — CHANGES_REQUESTED** [review]
I'm blocking this until we all get a chance to discuss as a team.

Really, really love this idea. I can think of four different problems this API could solve right off the bat—cache configuration, fragment/partial support, streaming control, and ISR. 

That's very exciting, but it's also a good sign that we shouldn't rush to merge this without going through [the roadmap process](https://github.com/withastro/roadmap). We have a process in place precisely because PRs like this can have wide-ranging implications and we owe it to our users to think through all the consequences of them before merging.

**Comment by ottomated** [comment]
> I'm blocking this until we all get a chance to discuss as a team.

Sounds good! I don't have time to put together an RFC, so I'll back out and leave this code to be used as it can!

**Comment by natemoo-re** [comment]
> Sounds good! I don't have time to put together an RFC, so I'll back out and leave this code to be used as it can!

No worries at all, that's totally understandable! Hoping we can pick this up ASAP to explore for 3.1 or soon after.

**Comment by sevfurneaux** [comment]
This is great @ottomated! @lilnasy pointed me to this PR, as it'll be a precursor to potential page specific adaptor configuration (ISR, Vercel On-Demand ISR, Vercel Draft mode, Netlify Builders for singular pages) as outlined in [this proposal](https://github.com/withastro/roadmap/discussions/656).

One thing worth considering, is allowing for object properties to be exportable, and not just booleans. This would allow for more advanced configurations such as ISR expiry time and other adaptor specific configurations as mentioned in [the proposal](https://github.com/withastro/roadmap/discussions/656):

```js
export const config = {
  isr: {
    expiration: 60,
    bypassToken: BYPASS_TOKEN,
    allowQuery: ['search']
  }
};
```

Exportable objects would also allow for more granular cache configuration, fragment/partial support, streaming control as you mentioned @natemoo-re.

**Comment by matthewp** [comment]
Still really like this idea. Hope it can get rolled into an RFC in the future. Closing as there's nothing actionable to be done right now.

**Pull request closed without merging**
